python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -61,24 +60,19 @@ cdef class Scalar:
             A NumPy dtype.
         """
 
-        value = cudf.utils.dtypes.to_cudf_compatible_scalar(value, dtype=dtype)
-
-        valid = value is not None
-
-        if dtype is None:
-            if value is None:
-                raise TypeError(
-                    "dtype required when constructing a null scalar"
-                )
-            else:
-                dtype = value.dtype
+        # Caching Mechanism
+        self._set_device_value(value, dtype)
 
```

Comment:
Minor but: the name `DeviceScalar._set_device_value()` is twice qualified. Maybe just `DeviceScalar._set_value()`? Ditto for `get`.